File: DSA_Level-1/Recursion and Backtracking/Recursion with Arraylists/Get Maze Paths/GetMazePaths_Solution.java

import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        Scanner scn=new Scanner(System.in);
        int n=scn.nextInt();
        int m=scn.nextInt();
        ArrayList<String> res=getMazePaths(0,0,n-1,m-1);
        System.out.println(res);
    }

    // sr - source row
    // sc - source column
    // dr - destination row
    // dc - destination column
    public static ArrayList<String> getMazePaths(int sr, int sc, int dr, int dc) {
        
        if(sr==dr && sc==dc){
            ArrayList<String> res=new ArrayList<>();
            res.add("");
            return res;
        }
        ArrayList<String> path1=new ArrayList<>();
        ArrayList<String> path2=new ArrayList<>();
        
        ArrayList<String> paths=new ArrayList<>();
        
        if(sr<dr){
            path1=getMazePaths(sr+1,sc,dr,dc); //Vertical moment
        }
        
        if(sc<dc){
            path2=getMazePaths(sr,sc+1,dr,dc); //horizontal moment
        }
        
        for(String s:path2){
            paths.add("h"+s);
        }
        
        for(String s:path1){
            paths.add("v"+s);
        }
        
        return paths;
    }

}

File: DSA_Level-1/Recursion and Backtracking/Recursion with Arraylists/Get Maze Paths/GetMazePaths_Solution.js

function mazePaths(sr,sc,dr,dc){

    if(sr==dr-1 && sc==dc-1){
        let res=[];
        res.push("");
        return res;
    }

    let pathH=[];
    let pathV=[];

    let paths=[];

    if(sr<dr-1){
        pathH=mazePaths(sr+1,sc,dr,dc); //horizontal movement
    }

    if(sc<dc-1){
        pathV=mazePaths(sr,sc+1,dr,dc); //Vertical movement
    }

    for(let path of pathH){
        paths.push("h"+path);
    }

    for(let path of pathV){
        paths.push("v"+path);
    }

    return paths;
}

console.log(mazePaths(0,0,3,3));